File: days/day3/solution.js
export default {
  a: inputs => {
    // Literally make a grid of 1000*1000 and fill it with dots
    const grid = Array(1000)
    for (let index = 0; index < grid.length; index++) {
      grid[index] = Array(1000).fill('.')
    }
    let count = 0
    grid[1][1] = 'x'

    // For every instruction, parse out into bits that are needed
    inputs.forEach(input => {
      const instruction = input.split(' ')
      const coords = instruction[2].replace(':', '').split(',')
      const size = instruction[3].split('x')

      // Loop through each instruction and add a `1` in every square that should be placed
      // If an overlap happens, put an x and increase the count of overlaps
      for (let row = 0; row < size[0]; row++) {
        for (let column = 0; column < size[1]; column++) {
          const rowVal = parseInt(coords[0]) + row
          const colVal = parseInt(coords[1]) + column
          const gridValue = grid[rowVal][colVal]

          if (gridValue === '.') {
            grid[rowVal][colVal] = 1
          } else if (gridValue === 1) {
            grid[rowVal][colVal] = 'x'
            count++
          }
        }
      }
    })

    // Return overlaps
    return count
  },
  b: inputs => {
    // All this is exactly the same as part a
    const grid = Array(1000)
    for (let index = 0; index < grid.length; index++) {
      grid[index] = Array(1000).fill('.')
    }
    let retVal = null
    grid[1][1] = 'x'

    inputs.forEach(input => {
      const instruction = input.split(' ')
      const coords = instruction[2].replace(':', '').split(',')
      const size = instruction[3].split('x')

      for (let row = 0; row < size[0]; row++) {
        for (let column = 0; column < size[1]; column++) {
          const rowVal = parseInt(coords[0]) + row
          const colVal = parseInt(coords[1]) + column
          const gridValue = grid[rowVal][colVal]

          if (gridValue === '.') {
            grid[rowVal][colVal] = 1
          } else if (gridValue === 1) {
            grid[rowVal][colVal] = 'x'
          }
        }
      }
    })

    // _THIS_ is just genious /s
    // This loops through all the instructions _backwards_ after having already
    // gone through the instructions in the original order
    // that way I don't need to worry about whether or not a future instruction
    // doesn't overlap with this one, because all the future instructions have been run already
    for (let index = inputs.length - 1; index > 0; index--) {
      const input = inputs[index]
      const instruction = input.split(' ')
      const coords = instruction[2].replace(':', '').split(',')
      const size = instruction[3].split('x')
      let overLapSize = 0

      for (let row = 0; row < size[0]; row++) {
        for (let column = 0; column < size[1]; column++) {
          const rowVal = parseInt(coords[0]) + row
          const colVal = parseInt(coords[1]) + column
          const gridValue = grid[rowVal][colVal]

          if (gridValue === 'x') {
            overLapSize++
          }
        }
      }

      if (overLapSize === 0) {
        retVal = instruction[0]
      }
    }

    return retVal
  }
}

File: days/day3/solution.py
import itertools
from pathlib import Path

p = Path('inputFile.txt')
fileLines = p.read_text().splitlines()

tests = {
    'a': [
        {'input': ['#1 @ 1,3: 4x4', '#2 @ 3,1: 4x4',
                   '#3 @ 5,5: 2x2'], 'expected': 4}

    ],
    'b': [
        {'input': ['#1 @ 1,3: 4x4', '#2 @ 3,1: 4x4',
                   '#3 @ 5,5: 2x2'], 'expected': '#3'}
    ]}


def part_a(inputs):
    if len(inputs) == 3:
        rangeSize = 10
    else:
        rangeSize = 1000

    count = 0

    grid = [['.'] * rangeSize for _ in range(rangeSize)]

    for input in inputs:
        instruction = input.split(' ')
        coords = instruction[2].replace(':', '').split(',')
        size = instruction[3].split('x')

        for row in range(int(size[0])):
            for col in range(int(size[1])):
                rowVal = int(coords[0]) + row
                colVal = int(coords[1]) + col
                gridValue = grid[rowVal][colVal]

                if gridValue == '.':
                    grid[rowVal][colVal] = 1
                elif gridValue == 1:
                    grid[rowVal][colVal] = 'x'
                    count = count + 1

    return count


def part_b(inputs):
    if len(inputs) == 3:
        rangeSize = 10
    else:
        rangeSize = 1000

    count = 0

    grid = [['.'] * rangeSize for _ in range(rangeSize)]

    for input in inputs:
        instruction = input.split(' ')
        coords = instruction[2].replace(':', '').split(',')
        size = instruction[3].split('x')

        for row in range(int(size[0])):
            for col in range(int(size[1])):
                rowVal = int(coords[0]) + row
                colVal = int(coords[1]) + col
                gridValue = grid[rowVal][colVal]

                if gridValue == '.':
                    grid[rowVal][colVal] = 1
                elif gridValue == 1:
                    grid[rowVal][colVal] = 'x'
                    count = count + 1

    for index in range(len(inputs), 0, -1):
        input = inputs[index - 1]
        instruction = input.split(' ')
        coords = instruction[2].replace(':', '').split(',')
        size = instruction[3].split('x')
        overlapSize = 0

        for row in range(int(size[0])):
            for col in range(int(size[1])):
                rowVal = int(coords[0]) + row
                colVal = int(coords[1]) + col
                gridValue = grid[rowVal][colVal]

                if gridValue == 'x':
                    overlapSize = overlapSize + 1

        if overlapSize == 0:
            return instruction[0]

    return count


solutions = {
    'a': part_a,
    'b': part_b
}


def test(part):
    for index in range(len(tests[part])):
        test = tests[part][index]
        result = solutions[part](test['input'])
        if result == test['expected']:
            print(f"Pass {index + 1}.")
        else:
            print(
                f"Fail {index + 1}: Expected {test['expected']}  got {result}")


def process(part):
    print(solutions[part](fileLines))


test('a')
process('a')

test('b')
process('b')
